src/charm.py

```diff
@@ -19,30 +21,48 @@
 class MysqlOperatorCharm(CharmBase):
     """A Juju Charm for MySQL InnoDB Cluster."""
 
+    _stored = StoredState()
+
     def __init__(self, *args):
         super().__init__(*args)
 
-        self.framework.observe(self.on.start, self._on_start)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        change_events = [
+        self._src_dir = Path(__file__).parent
+        self._server_container = "mysql-server"
+        self._router_container = "mysql-router"
+        self._stored.set_default(root_password="", mysql_password="")
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        pebble_events = [
             self.on.mysql_server_pebble_ready,
             self.on.mysql_router_pebble_ready,
         ]
-        for event in change_events:
+        for event in pebble_events:
             self.framework.observe(event, self._on_pebble_ready)
 
     # ---- Event handlers
-    def _on_start(self, _):
-        """Placeholder for StartEvent."""
-        pass
+    def _on_install(self, event: WorkloadEvent) -> None:
+        """Ensures the MySQL server instance is configured for an InnoDB Cluster."""
+        container = self.unit.get_container(self._server_container)
+        if not container.can_connect():
+            event.defer()
+            return
```

Comment:
It looks like you're doing most of your install tasks during your _pebble_ready handler. I think that you can simply leave the _on_install handler out of this charm.